fastapi/dependencies/utils.py

```diff
@@ -433,184 +516,217 @@ def is_gen_callable(call: Callable[..., Any]) -> bool:
+
+    # TODO make async task
+    for node_idx, body_params in plan.body_params_list:
         (
             body_values,
             body_errors,
         ) = await request_body_to_args(  # body_params checked above
-            required_params=dependant.body_params, received_body=body
+            required_params=body_params, received_body=body
         )
-        values.update(body_values)
+        nodes[node_idx].arguments.update(body_values)
         errors.extend(body_errors)
-    if dependant.http_connection_param_name:
-        values[dependant.http_connection_param_name] = request
-    if dependant.request_param_name and isinstance(request, Request):
-        values[dependant.request_param_name] = request
-    elif dependant.websocket_param_name and isinstance(request, WebSocket):
-        values[dependant.websocket_param_name] = request
-    if dependant.background_tasks_param_name:
-        if background_tasks is None:
-            background_tasks = BackgroundTasks()
-        values[dependant.background_tasks_param_name] = background_tasks
-    if dependant.response_param_name:
-        values[dependant.response_param_name] = response
-    if dependant.security_scopes_param_name:
-        values[dependant.security_scopes_param_name] = SecurityScopes(
-            scopes=dependant.security_scopes
-        )
-    return values, errors, background_tasks, response, dependency_cache
 
+    if errors:
+        return nodes[0].result, errors, background_tasks, response
 
-def request_params_to_args(
-    required_params: Sequence[ModelField],
-    received_params: Union[Mapping[str, Any], QueryParams, Headers],
-) -> Tuple[Dict[str, Any], List[ErrorWrapper]]:
-    values = {}
-    errors = []
-    for field in required_params:
-        if is_scalar_sequence_field(field) and isinstance(
-            received_params, (QueryParams, Headers)
-        ):
-            value = received_params.getlist(field.alias) or field.default
-        else:
-            value = received_params.get(field.alias)
-        field_info = field.field_info
-        assert isinstance(
-            field_info, params.Param
-        ), "Params must be subclasses of Param"
-        if value is None:
-            if field.required:
-                errors.append(
-                    ErrorWrapper(
-                        MissingError(), loc=(field_info.in_.value, field.alias)
-                    )
-                )
-            else:
-                values[field.name] = deepcopy(field.default)
-            continue
-        v_, errors_ = field.validate(
-            value, values, loc=(field_info.in_.value, field.alias)
-        )
-        if isinstance(errors_, ErrorWrapper):
-            errors.append(errors_)
-        elif isinstance(errors_, list):
-            errors.extend(errors_)
-        else:
-            values[field.name] = v_
-    return values, errors
+    async with anyio.create_task_group() as task_group:
```

Comment:
Thanks for benchmarking. I've made some optimization for simple cases.
Also i added special case for sync functions, to avoid threadpool execution (for simple cases without IO).
Here is your benchmark example updated:
```python
from fastapi import FastAPI, Depends
from fastapi.testclient import TestClient
try:
    from fastapi.dependencies.utils import sync_dependency
except ImportError:
    def sync_dependency(f):
        return f

app = FastAPI()

Dep0 = Depends(sync_dependency(lambda: "dep0"))
Dep1 = Depends(sync_dependency(lambda: "dep1"))
Dep2 = Depends(sync_dependency(lambda: "dep2"))
Dep3 = Depends(sync_dependency(lambda: "dep3"))
Dep4 = Depends(sync_dependency(lambda: "dep4"))
Dep5 = Depends(sync_dependency(lambda: "dep5"))
Dep6 = Depends(sync_dependency(lambda: "dep6"))
Dep7 = Depends(sync_dependency(lambda: "dep7"))
Dep8 = Depends(sync_dependency(lambda: "dep8"))
Dep9 = Depends(sync_dependency(lambda: "dep9"))


@app.get("/")
async def read_main(
    dep0: str = Dep0,
    dep1: str = Dep1,
    dep2: str = Dep2,
    dep3: str = Dep3,
    dep4: str = Dep4,
    dep5: str = Dep5,
    dep6: str = Dep6,
    dep7: str = Dep7,
    dep8: str = Dep8,
    dep9: str = Dep9
):
    return {"msg": "Hello World"}


client = TestClient(app)


def test_read_main():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {"msg": "Hello World"}


if __name__ == '__main__':
    import timeit
    print(timeit.timeit("test_read_main()", globals=globals(), number=3000))
    
# mine: 3.93
# 0.70.0: 8.98
# 0.69.0: 8.98
# 0.68.2: 9.06
```
I got 2x performance gain for synthetic case :)